sljit_src/sljitWXExecAllocator.c

```diff
@@ -0,0 +1,80 @@
+/*
+ *    Stack-less Just-In-Time compiler
+ *
+ *    Copyright Zoltan Herczeg (hzmester@freemail.hu). All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are
+ * permitted provided that the following conditions are met:
+ *
+ *   1. Redistributions of source code must retain the above copyright notice, this list of
+ *      conditions and the following disclaimer.
+ *
+ *   2. Redistributions in binary form must reproduce the above copyright notice, this list
+ *      of conditions and the following disclaimer in the documentation and/or other materials
+ *      provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) AND CONTRIBUTORS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER(S) OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+   This file contains a simple W^X executable memory allocator for POSIX systems.
+
+   It allocates a separate block for each code block which may waste a lot of memory.
+*/
+
+#include <sys/mman.h>
+#include <unistd.h>
+
+SLJIT_API_FUNC_ATTRIBUTE void* sljit_malloc_exec(sljit_uw size)
+{
+	sljit_uw* ptr;
+
+	size += sizeof(sljit_uw);
```

Comment:
I think we should use here `max_align_t` to ensure that the data/code is properly aligned.

Reply:
Alternatively... can we keep the list of allocations somewhere else in a dedicated structure?

Reply:
Something similar is done in the RW+RX allocator, that we allocate:

```
struct chunk_header {
        void *executable;
};
```

We could go here for mmap()/malloc()ing:

```
struct chunk_header {
        void *address;
        sljit_uw size;
};
```

and then mmap PAGE_SIZE-rounded chunks.

Reply:
@krytarowski: max_align_t is C11 and so we will need to add compatibility for older compilers or drop support for little gain IMHO

I was using instead size_t which should be enough in practice (testing it works in all architectures NetBSD supports would help us to confirm that, as mentioned before), but considering sljit_uw is used already in sljit where size_t would had been used I see why this made sense here as well.

do you have any reason to believe some architectures will have problems with code aligned to 8 (in 64-bit) or 4 (in 32-bit) bytes and might need 16 byte alignment from the base of the page?, note that the code ptr is not accessed in ways that could cause alignment issues AFAIK

Reply:
@carenas It depends on code generation. max_align_t is certainly safer and it can be easily implemented:

```
typedef union {
        void *_v;
        long double _ld;
        long long int _ll;
} sljit_max_align_t;

size += sizeof(sljit_max_align_t);
```